File: web/src/api/solutions.ts
import { apiGet, buildQuery } from './client'

// Mirrors internal/models.Solution. project is absent for a global gotcha.
export interface Solution {
  id: string
  project?: string
  error_description: string
  solution: string
  tags: string[]
  source_url: string
  created_at: string
  updated_at: string
}

export interface SolutionFilter {
  project?: string
  tag?: string
}

export function listSolutions(f: SolutionFilter = {}): Promise<Solution[]> {
  return apiGet<{ items: Solution[] }>(
    `/api/v1/solutions${buildQuery({ project: f.project, tag: f.tag })}`,
  ).then((r) => r.items)
}

File: internal/api/solutions.go
package api

import (
	"net/http"

	"github.com/jeroenpf/mneme/internal/models"
	"github.com/jeroenpf/mneme/internal/store"
)

// SolutionsHandler serves the read-only error/solution REST surface. Writes
// go through the MCP log_solution tool, and ranked search through
// find_solution — REST is filter-only, not search.
type SolutionsHandler struct{ Store store.Store }

type solutionListResponse struct {
	Items []*models.Solution `json:"items"`
}

func (h *SolutionsHandler) List(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	var f store.SolutionFilter
	if p := q.Get("project"); p != "" {
		f.Project = &p
	}
	if tag := q.Get("tag"); tag != "" {
		f.Tags = []string{tag}
	}
	items, err := h.Store.ListSolutions(r.Context(), f)
	if err != nil {
		writeStoreError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, solutionListResponse{Items: items})
}
